pinot-server/src/main/java/org/apache/pinot/server/starter/helix/BaseServerStarter.java

```diff
@@ -501,6 +502,16 @@ private void startupServiceStatusCheck(long endTimeMs) {
       }
     }
 
+    boolean exitServerOnIncompleteStartup = _serverConf.getProperty(Server.CONFIG_OF_EXIT_SERVER_ON_INCOMPLETE_STARTUP,
+        Server.DEFAULT_EXIT_SERVER_ON_INCOMPLETE_STARTUP);
+    if (exitServerOnIncompleteStartup) {
+      String errorMessage = String.format("Service status %s has not turned GOOD within %dms: %s. Exiting server.",
+          serviceStatus, System.currentTimeMillis() - startTimeMs, ServiceStatus.getStatusDescription());
+      LOGGER.error(errorMessage);
+      // Stop the server so that it will be removed from the Helix cluster
+      stop();
```

Comment:
Suggest not calling `stop()` here since the initialization is not done yet. It should be enough to just:
```
  _adminApiApplication.stop();
  _helixManager.disconnect();
```

Reply:
that's fair